include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -884,6 +888,178 @@ namespace VectorTools
+
+    constraints.distribute (vec);
+
+    const IndexSet locally_owned_dofs = dof.locally_owned_dofs();
+    IndexSet::ElementIterator it = locally_owned_dofs.begin();
+    for (; it!=locally_owned_dofs.end(); ++it)
+      vec_result(*it) = vec(*it);
+    vec_result.compress(VectorOperation::insert);
+  }
+
+
+  template <int dim, typename VectorType, int fe_degree, int n_q_points_1d>
+  void project_parallel (const MatrixFree<dim,typename VectorType::value_type> &matrix_free,
+                         const ConstraintMatrix &constraints,
+                         const std::function< VectorizedArray<typename VectorType::value_type> (const unsigned int, const unsigned int)> func,
+                         VectorType &vec_result)
+  {
+    const DoFHandler<dim> &dof = matrix_free.get_dof_handler();
+
+    typedef typename VectorType::value_type number;
+    Assert (dof.get_fe().n_components() == 1,
+            ExcDimensionMismatch(dof.get_fe().n_components(),
+                                 1));
+    Assert (vec_result.size() == dof.n_dofs(),
+            ExcDimensionMismatch (vec_result.size(), dof.n_dofs()));
+    Assert (dof.get_fe().degree == fe_degree,
+            ExcDimensionMismatch(fe_degree, dof.get_fe().degree));
+
+    typedef MatrixFreeOperators::MassOperator<dim, fe_degree, 1, number> MatrixType;
+    MatrixType mass_matrix;
+    mass_matrix.initialize(matrix_free);
+    mass_matrix.compute_diagonal();
+
+    typedef LinearAlgebra::distributed::Vector<number> LocalVectorType;
+    LocalVectorType vec, rhs, inhomogeneities;
+    matrix_free.initialize_dof_vector(vec);
+    matrix_free.initialize_dof_vector(rhs);
+    matrix_free.initialize_dof_vector(inhomogeneities);
+    constraints.distribute(inhomogeneities);
+    inhomogeneities*=-1.;
+
+    // assemble right hand side:
+    {
+      FEEvaluation<dim,fe_degree,n_q_points_1d,1,number> fe_eval(matrix_free);
+      const unsigned int n_cells = matrix_free.n_macro_cells();
+      const unsigned int n_q_points = fe_eval.n_q_points;
+
+      for (unsigned int cell=0; cell<n_cells; ++cell)
+        {
+          fe_eval.reinit(cell);
+          for (unsigned int q= 0; q < n_q_points; ++q)
+            fe_eval.submit_value(func(cell,q), q);
+
+          fe_eval.integrate (true,false);
+          fe_eval.distribute_local_to_global (rhs);
+        }
+      rhs.compress(VectorOperation::add);
+    }
+
+    mass_matrix.vmult_add(rhs, inhomogeneities);
+
+    //now invert the matrix
+    ReductionControl     control(5*rhs.size(), 0., 1e-12, false, false);
+    SolverCG<LinearAlgebra::distributed::Vector<number> > cg(control);
+    typename PreconditionChebyshev<MatrixType, LocalVectorType>::AdditionalData data;
+    data.matrix_diagonal_inverse = mass_matrix.get_matrix_diagonal_inverse();
+    PreconditionChebyshev<MatrixType, LocalVectorType> preconditioner;
+    preconditioner.initialize(mass_matrix, data);
+    cg.solve (mass_matrix, vec, rhs, preconditioner);
```

Comment:
Out of curiosity, do you happen to know how many iterations this typically takes? I'm thinking that it may actually be interesting for ASPECT to use the matrix-free idea to invert mass matrices :-)